### PR TITLE
introduce GetImageWithDetails to manage image inspect data in backend

### PR DESCRIPTION
**Pull request by ndeloof**
**- What I did**
Introduce GetImageWithDetails so that "image inspect" logic is not implement by router but backend, and we can re-implement using containerd snapshotter

relates to https://github.com/moby/moby/pull/43680

### CODE REVIEW

**Review by rumpl — APPROVED** [review]

File: api/server/router/image/backend.go [hunk 1]
```diff
@@ -25,6 +25,7 @@ type imageBackend interface {
 	ImageHistory(imageName string) ([]*image.HistoryResponseItem, error)
 	Images(ctx context.Context, opts types.ImageListOptions) ([]*types.ImageSummary, error)
 	GetImage(ctx context.Context, refOrID string, platform *specs.Platform) (*dockerimage.Image, error)
+	GetImageWithDetails(ctx context.Context, refOrID string, platform *specs.Platform) (*dockerimage.Image, *dockerimage.Details, error)
```
[the remark this answers is not in the document]
Comment (ndeloof) [reply api/server/router/image/backend.go:28]: I can give it a try

[the remark this answers is not in the document]
Comment (ndeloof) [reply api/server/router/image/backend.go]: implemented as [`bda80dc` (#27)](https://github.com/rumpl/moby/pull/27/commits/bda80dcd6506ce982fe4a87ee70fedc8b92fa9d7)
side effect: `Backend` interface, as declared by `image` package in router, now depends on `daemon.GetImageOpts` type where this functional option is defined, which tends to demonstrate those `Backend` interface are fully artificial in the existing design

[the remark this answers is not in the document]
Comment (ndeloof) [reply api/server/router/image/backend.go]: right, but https://github.com/moby/moby/pull/43680 initial intent was to avoid the backend to return API types, imho we should then also _not_ use those as inputs.

File: image/image.go [hunk 2]
```diff
@@ -112,6 +112,17 @@ type Image struct {
...
+	// Details holds additional details about image
+	Details *Details
+}
+
+// Details provides additional image data
+type Details struct {
+	Size        int64
+	Metadata    map[string]string
+	Driver      string
+	LastUpdated time.Time
```
Comment (thaJeztah) [inline image/image.go:125]: For the new fields, make sure they are not serialised to JSON (`json:"-"`)

not sure if it works to add that to `Details`, or it if's needed for the individual fields (or _both_) (haven't tested)

File: api/server/router/image/backend.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thaJeztah) [reply api/server/router/image/backend.go]: They have to be connected "somewhere". And I'd say it's better to keep the daemon out of the API (there may actually be a check somewhere for that)

[the remark this answers is not in the document]
Comment (ndeloof) [reply api/server/router/image/backend.go]: > better to keep the daemon out of the API 

but then we get the API inside the daemon, while the `backend` interfaces _should_ enforce separation. imho API should only rely on engine types, but that's a bigger refactoring :P

**Comment by ndeloof** [comment]
`import cycle not allowed` :'(

File: image/image.go (hunk 2 above)
Comment (ndeloof) [reply image/image.go:125]: makes me wonder: why do we define json rules for an engine type? This one is not part of the API

File: api/server/router/image/backend.go [hunk 3]
```diff
@@ -24,7 +24,7 @@ type imageBackend interface {
 	ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error)
 	ImageHistory(imageName string) ([]*image.HistoryResponseItem, error)
 	Images(ctx context.Context, opts types.ImageListOptions) ([]*types.ImageSummary, error)
-	GetImage(ctx context.Context, refOrID string, platform *specs.Platform) (*dockerimage.Image, error)
+	GetImage(ctx context.Context, refOrID string, platform *specs.Platform, opts ...image.GetImageOptsFn) (*dockerimage.Image, error)
```
Comment (thaJeztah) [inline api/server/router/image/backend.go:27]: One thing that caught my eye, and we probably should discuss; going forward do we want _functional arguments_ (`WithOptionFoo()`) or option-structs? (`FooOptions{}`)

I know we're a bit inconsistent in this already (although so far, differences were between packages, but (mostly?) within the same package we stuck to the same approach.

Mostly asking here, because in this case we start to mix approaches;

- `Image` takes `types.ImageListOptions`, which is a struct
- `GetImage` takes `image.GetImageOptsFn`, which is a functional argument (nit: probably the `Fn` is redundant)

File: daemon/containerd/service.go [hunk 4]
```diff
@@ -569,34 +569,86 @@ func (cs *containerdStore) CommitImage(c backend.CommitConfig) (image.ID, error)
 	panic("not implemented")
 }
 
-func (cs *containerdStore) GetImage(ctx context.Context, refOrID string, platform *v1.Platform) (*image.Image, error) {
-	desc, err := cs.ResolveImage(ctx, refOrID)
+// GetImageWithDetails returns an image corresponding to the image referred to by refOrID and additional details about layer.
+func (cs *containerdStore) GetImageWithDetails(ctx context.Context, refOrID string, platform *v1.Platform) (*image.Image, *image.Details, error) {
```
Comment (thaJeztah) [inline daemon/containerd/service.go:573]: Was this one still needed? Or can be reverted as part of the last changes?

File: api/server/router/image/backend.go (hunk 3 above)
Comment (ndeloof) [reply api/server/router/image/backend.go:27]: the reason I used this was to avoid introducing a parameter that would be `nil` for all callers but the "inspect" use-case

File: image/image.go [hunk 5]
```diff
@@ -112,6 +112,17 @@ type Image struct {
...
 	// Not to be confused with the legacy V1 ID in V1Image.
 	computedID ID
+
+	// Details holds additional details about image
+	Details *Details `json:"-"`
+}
+
+// Details provides additional image data
+type Details struct {
+	Size        int64
```
Comment (thaJeztah) [inline image/image.go:122]: The only thing I'm still looking at is how this Size relates to the `V1Image.Size` field defined higher up, which is embedded in the Image type; https://github.com/moby/moby/blob/3c6c9fa180eb189636dbee5ca608816eb6bcd82c/image/image.go#L84-L85


Reason I was looking at that, was that if that field could be used, we could embed the `Details` struct, so that the additional fields would become "part" of the image; admitted, slightly more cumbersome to _create_, but more convenient to "read" the properties; https://go.dev/play/p/y6F5NI-RHXo

File: image/image.go [hunk 6]
```diff
@@ -112,6 +112,17 @@ type Image struct {
...
+	Metadata    map[string]string
```
Comment (thaJeztah) [inline image/image.go:123]: Just realised that this is to store _graphdriver_ metadata; wondering if that's clear enough from this type (at least could use a GoDoc), or was the intent to re-purpose this field for something else following the containerd integration?

File: api/server/router/image/backend.go [hunk 7]
```diff
@@ -22,9 +22,9 @@ type Backend interface {
 
 type imageBackend interface {
 	ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error)
-	ImageHistory(imageName string) ([]*image.HistoryResponseItem, error)
+	ImageHistory(ctx context.Context, imageName string) ([]*image.HistoryResponseItem, error)
```
Comment (thaJeztah) [inline api/server/router/image/backend.go:25]: Unrelated change?

Comment (thaJeztah) [reply api/server/router/image/backend.go:25]: Ah, looks like it's because it's calling `GetImage` internally; alternatively you could put a `context.TODO()` in those if more changes are needed (as we're already passing `nil` as context)

File: daemon/images/image_history.go [hunk 8]
```diff
@@ -11,9 +12,9 @@ import (
 
 // ImageHistory returns a slice of ImageHistory structures for the specified image
 // name by walking the image lineage.
-func (i *ImageService) ImageHistory(name string) ([]*image.HistoryResponseItem, error) {
+func (i *ImageService) ImageHistory(ctx context.Context, name string) ([]*image.HistoryResponseItem, error) {
 	start := time.Now()
-	img, err := i.GetImage(nil, name, nil)
+	img, err := i.GetImage(nil, name, image.GetImageOpts{})
```
Comment (thaJeztah) [inline daemon/images/image_history.go:17]: Does this need to pass the context? Instead of `nil`? (Alternatively, set `context.TODO()` if it needs work / looked at separately)

File: image/image.go (hunk 6 above)
Comment (ndeloof) [reply image/image.go:123]: I actually wonder why graphdriver metadata are exposed to the API, seems like debugging data as a user would hardly use image's `mergedir`

File: image/image.go (hunk 5 above)
Comment (ndeloof) [reply image/image.go:122]: `Size` vs `VirtualSize` is a mess. Should have been deprecated when code evolved for both to return same value.

File: api/server/router/image/backend.go (hunk 7 above)
Comment (ndeloof) [reply api/server/router/image/backend.go:25]: There's too much TODO already :P

File: daemon/images/image_history.go (hunk 8 above)
Comment (ndeloof) [reply daemon/images/image_history.go:17]: I just missed this one

File: image/image.go (hunk 5 above)
Comment (ndeloof) [reply image/image.go:122]: doc says "Size is the total size of the image including all layers it is composed of" which means this is == VirtualSize
but I can't find any place this value is set 🤨

File: image/image.go (hunk 2 above)
Comment (thaJeztah) [reply image/image.go]: I _think_ (but would have to double-check) these types are also used to serialise image information to disk (and as such could be part of, e.g., save/load. So in that case it's not for the _API_, but to prevent persisting this data to disk.

IIRC, this is also the reason why the `V1Image` is embedded, as that's what's used for save/load

File: image/image.go (hunk 6 above)
Comment (thaJeztah) [reply image/image.go:123]: Yes, I agree. It looks like this was one of the Red Hat contributions, and I can only "guess" how they wanted to use this (possibly as part of some systemd hackery); https://github.com/moby/moby/pull/13198 (also related; https://github.com/moby/moby/pull/13130)

File: image/image.go (hunk 5 above)
Comment (thaJeztah) [reply image/image.go:122]: Yes, looks like a TODO was added (but in the wrong place, not on the definition of the type) when the migration to content-addressable store was done; https://github.com/moby/moby/blob/e3079b470494a596bc8b35bc682e7c82ac9f0f9f/daemon/daemon.go#L1081

Before that, Size and VirtualSize were different things; https://github.com/moby/moby/blob/694df3ff9f0f6b6acd4f3d866d443c740cce3f3f/graph/service.go#L54-L55

Comment (thaJeztah) [reply image/image.go]: Let's keep this as-is for now; as they're basically internal (type isn't directly exposed on the API, and the `json:"-"` should prevent them from leaking to files (or API), we can still change things if we want to.
